include/glow/Backends/DeviceManager.h

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_BACKENDS_DEVICEMANAGER_H
+#define GLOW_BACKENDS_DEVICEMANAGER_H
+
+#include "glow/Backends/Backend.h"
+#include "glow/Backends/CompiledFunction.h"
+#include "glow/Graph/Context.h"
+#include "glow/Graph/Graph.h"
+#include "glow/Support/ThreadPool.h"
+
+#include <atomic>
+#include <condition_variable>
+#include <map>
+#include <mutex>
+#include <thread>
+#include <unordered_map>
+
+namespace glow {
+
+using DeviceNetworkID = size_t;
+
+enum ResultCode { READY, EXECUTED, FAILED, CANCELLED };
+
+using ReadyCB = std::function<void(const Module *, ResultCode)>;
+using ResultCB = std::function<void(ResultCode, std::unique_ptr<Context>)>;
+using FunctionMap = std::map<std::string, CompiledFunction *>;
+
+class DeviceManager {
+protected:
+  // Type of Backend for this Device.
+  BackendKind backend_;
+  /// Thread which interfaces with the device.
+  ThreadPool workThread_;
+
+public:
+  DeviceManager(BackendKind backend);
+  virtual ~DeviceManager();
+
+  /// Initialize the device.
+  void init();
+
+  /// Load the provided module into the device, readyCB will be called when
+  /// ready to use
+  void addNetwork(const Module *module, FunctionMap functions, ReadyCB readyCB);
+
+  /// Remove (and delete) the provided network and all it's functions, freeing
+  /// up space on the device.
+  void evictNetwork(const Module *module);
+
+  /// Execute the named Function in an already provided network on the device.
+  /// functionName must match the name of a function already added.
+  /// Context should have all Placeholders allocated. resultCB will be called
+  /// with the Context results filled.
+  void runFunction(std::string functionName, std::unique_ptr<Context> ctx,
```

Comment:
Is it possible that the same function is started with different contexts and we have multiple running instances of the function? Do we need to return some kind of unique id from `runFunction`, so that we can distinguish between multiple running instances? This could be useful e.g. for cancellation, profiling, monitoring the execution, etc.

Reply:
To a certain degree the Context pointer itself is that identifier, but if we want something more specific I figured we could put it into the lambda capture of the callback (e.g. how we use it in the tests).

How do you picture using it and the id?

Reply:
My idea was that if we associate an id with the specific run, we could later query the device manager about the current state of the run by providing this id and ask how long this run took, when it was started, when it was finished, if it is still running, etc. We could also use it for cancellation, if we decide that we want to support cancellation.

And yes, I guess the Context pointer could be that id. But I'm not sure we want to guarantee it. Instead we could return an opaque id (which by default could be the Context pointer), which the client only uses to provide as input parameter to those query APIs I described above.

Reply:
> Do you think it's useful to pass it in to the callback as well?

Yes, it could be useful if we pass it to the callback. May be it wants to query some information about this specific run.